plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,534 @@
+ *
+ * The other mode is stateful, and uses redis to store the session state and prevent self-collisions when using
+ * $identify events.
+ *
+ * Stateless mode:
+ *
+ * The distinct id is a prefix + the base64 representation of the hash. The session ID is a UUIDv7, that uses some of
+ * the bytes of the hash as the random part, and the timestamp of the start of the day as the timestamp part.
+ *
+ * Stateful mode:
+ *
+ * The session ID is a UUIDv7 that's stored in redis, using the timestamp of the first event of the session as the
+ * timestamp part of the UUID. We implement our own session inactivity system rather than using redis TTLs, as the
+ * session inactivity period is 30 minutes, and it's not impossible for us to have well over 30 minutes of ingestion lag.
+ *
+ * To ensure that a user that logs in and out doesn't collide with themselves, we store the set of identify event UUIDs
+ * in redis against the base hash value, and use the number of identifies for that hash to calculate the final hash value.
+ * The exact number we append is the number of identifies that happened before this user called identify, which will
+ * mean adjusting the number based on whether the event was pre or post identify, or was itself the identify event.
+ */
+
+const TIMEZONE_FALLBACK = 'UTC'
+const COOKIELESS_SENTINEL_VALUE = '$posthog_cklsh'
+const COOKIELESS_MODE_FLAG_PROPERTY = '$cklsh_mode'
+const COOKIELESS_EXTRA_HASH_CONTENTS_PROPERTY = '$cklsh_extra'
+const MAX_NEGATIVE_TIMEZONE_HOURS = 12
+const MAX_POSITIVE_TIMEZONE_HOURS = 14
+const MAX_INGESTION_LAG_HOURS = 24
+const SALT_TTL_SECONDS =
+    (MAX_POSITIVE_TIMEZONE_HOURS + MAX_NEGATIVE_TIMEZONE_HOURS + MAX_INGESTION_LAG_HOURS + 24) * 60 * 60
+const SESSION_TTL_SECONDS = 60 * 60 * 24
+const IDENTIFIES_TTL_SECONDS = 60 * 60 * 24
+const DELETE_EXPIRED_SALTS_INTERVAL_MS = 60 * 60 * 1000
+
+export async function cookielessServerHashStep(
+    runner: EventPipelineRunner,
+    event: PluginEvent
+): Promise<[PluginEvent | undefined]> {
+    // if events aren't using this mode, skip all processing
+    if (!event.properties?.[COOKIELESS_MODE_FLAG_PROPERTY]) {
+        return [event]
+    }
+
+    // if the team isn't allowed to use this mode, drop the event
+    const team = await runner.hub.teamManager.getTeamForEvent(event)
+    if (!team?.cookieless_server_hash_mode) {
+        // TODO log
+        return [undefined]
+    }
+    const teamTimeZone = team.timezone
+
+    const timestamp = event.timestamp ?? event.sent_at ?? event.now
+
+    // drop some events that aren't valid in this mode
+    if (!timestamp) {
+        // TODO log
+        return [undefined]
+    }
+    const { $session_id: sessionId, $device_id: deviceId } = event.properties
+    if (sessionId != null || deviceId != null) {
+        // TODO log
+        return [undefined]
+    }
+
+    if (event.event === '$alias') {
+        // TODO support these
+        return [undefined]
+    }
```

Comment:
Do we even need to support these? Sounds kinda hard because we'll need to save a "canonical" id in Redis, right?

Reply:
Let's leave this for another day, we can go live without it